src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -30,124 +28,110 @@ internal static class DocumentBasedFixAllProviderHelpers
-
-            foreach (var fixAllContext in fixAllContexts)
-            {
-                Contract.ThrowIfFalse(
-                    fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project or FixAllScope.ContainingMember or FixAllScope.ContainingType);
-
-                // TODO: consider computing this in parallel.
-                var singleContextDocIdToNewRootOrText = await getFixedDocumentsAsync(fixAllContext, progressTracker).ConfigureAwait(false);
-
-                // Note: it is safe to blindly add the dictionary for a particular context to the full dictionary.  Each
-                // dictionary will only update documents within that context, and each context represents a distinct
-                // project, so these should all be distinct without collisions.  However, to be very safe, we use an
-                // overwriting policy here to ensure nothing causes any problems here.
-                foreach (var kvp in singleContextDocIdToNewRootOrText)
-                    allContextsDocIdToNewRootOrText[kvp.Key] = kvp.Value;
-            }
-        }
+            using var _ = await RemoteKeepAliveSession.CreateAsync(solution, cancellationToken).ConfigureAwait(false);
 
-        // Next, go and insert those all into the solution so all the docs in this particular project point at
-        // the new trees (or text).  At this point though, the trees have not been cleaned up.  We don't cleanup
-        // the documents as they are created, or one at a time as we add them, as that would cause us to run
-        // cleanup on N different solution forks (which would be very expensive).  Instead, by adding all the
-        // changed documents to one solution, and then cleaning *those* we only perform cleanup semantics on one
-        // forked solution.
-        var currentSolution = solution;
-        foreach (var (docId, (newRoot, newText)) in allContextsDocIdToNewRootOrText)
-        {
-            currentSolution = newRoot != null
-                ? currentSolution.WithDocumentSyntaxRoot(docId, newRoot)
-                : currentSolution.WithDocumentText(docId, newText!);
+            return await ProducerConsumer<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>>.RunParallelAsync(
+                source: fixAllContexts,
+                produceItems: static async (fixAllContext, callback, args, cancellationToken) =>
+                {
+                    using var _ = args.progressTracker.ItemCompletedScope();
+
+                    Contract.ThrowIfFalse(
+                        fixAllContext.Scope is FixAllScope.Document or FixAllScope.Project or FixAllScope.ContainingMember or FixAllScope.ContainingType);
+
+                    var singleContextDocIdToNewRootOrText = await args.getFixedDocumentsAsync(fixAllContext).ConfigureAwait(false);
+                    callback(singleContextDocIdToNewRootOrText);
```

Comment:
i'm likely going to change this to a callback as well, to avoid this intermediary dictionary.